README.md
# react-pretty-input

## Overview
A pretty input form field for React projects.

[![NPM version][npm-image]][npm-url]
[![Build Status][travis-badge]][travis-url]
[![Coverage Status][coverage-badge]][coverage-url]
[![npm download][download-image]][download-url]
[![MIT License][license-badge]][license-url]

## [Changelog](https://github.com/elmao79/react-pretty-input/blob/master/CHANGELOG.md)

## Installation
[![react-pretty-input](https://nodei.co/npm/react-pretty-input.png)](https://npmjs.org/package/react-pretty-input)

## Usage

ES6 import:
```js
import { PrettyInputText } from 'react-pretty-input';
```

```jsx
// In JSX
<PrettyInputText name="email" labelText="Email Address" errorValue="Invalid Email Address" isRequired />

// In vanilla JS
React.createElement(PrettyInputText, { name: 'email', labelText: 'Email Address', errorValue: 'Invalid Email Address', isRequired: true});
```

## Props

#### `name` (required) (type: `string`)
This prop is the name attribute of the generated input tag.

#### `labelText` (required) (type: `string`)
This prop is the description string of the field. It's used to show the placeholder and the label of the input to describe it.

#### `inputValue` (optional) (type: `string`)
This prop is the default value of the component.

#### `errorValue` (optional) (type: `string`) (default: `'Error'`)
This prop is the description string of and error occurred in the field. It's used to show the error in the label to describe it.

#### `size` (optional) (type: `integer`) (default: `50`)
This prop is used to limit the input string length of the field.

#### `isRequired` (optional) (type: `boolean`) (default: `false`)
This prop is used to indicate if the input string of the field is required. If the prop value is `true` and the field is empty an error is showed in label and border field.

#### `onValidation` (optional) (type: `function`) (default: `inputValue => true`)
This prop is used to pass a callback function to validate de input value of the component. This function must return a boolean value.

#### `onChange` (optional) (type: `function`) (default: `null`)
This prop is used to pass a callback function invoked on input text change.

#### `onKeyPress` (optional) (type: `function`) (default: `null`)
This prop is used to pass a callback function invoked on every key press event.

#### `labelColor` (optional) (type: `string`) (default: `#0069ff`)
This prop is used to set the color of the label.

#### `backgroundColor` (optional) (type: `string`) (default: `#fff`)
This prop is used to set the background color of the component.

#### `width` (optional) (type: `integer`) (default: `200`)
This prop is used to set the css width of the component.

## Contributing
Use [Github issues](https://github.com/elmao79/react-pretty-input/issues) for feature requests and bug reports. We actively accept pull requests.

## License

MIT

[npm-image]: https://img.shields.io/npm/v/react-pretty-input.svg?style=flat-square
[npm-url]: http://npmjs.org/package/react-pretty-input
[download-image]: https://img.shields.io/npm/dm/react-pretty-input.svg?style=flat-square
[download-url]: https://npmjs.org/package/react-pretty-input
[license-badge]: https://img.shields.io/npm/l/react-pretty-input.svg?style=flat-square
[license-url]: https://github.com/elmao79/react-pretty-input/blob/master/LICENSE
[coverage-badge]: https://coveralls.io/repos/github/elmao79/react-pretty-input/badge.svg?branch=master
[coverage-url]: https://coveralls.io/github/elmao79/react-pretty-input?branch=master
[travis-badge]: https://travis-ci.org/elmao79/react-pretty-input.svg?branch=master
[travis-url]: https://travis-ci.org/elmao79/react-pretty-input


CHANGELOG.md

## v1.0.9

### Added
- New prop `labelText` added to set the text of the label instead of `inputValue`.
- New prop `isEnabled` added to control whether the component is enabled or not.
- New prop `onChange` added to pass a callback function invoked on input change event.
- New prop `onKeyPress` added to pass a callback function invoked on every key press.
- New prop `labelColor` added to set the color of the component label.
- New prop `backgroundColor` added to set the background color of the component.
- New prop `width` added to set the width (in pixels) of the component.

### Changed
- Prop `inputValue` is not required now.

## v1.0.4

### Added
- New prop `onValidation` added. This prop is used to customize the validation function of the field.

### Changed
- Prop `errorValue` is not required now.

### Fixed
- Attribute `name` fixed for generated input.


src/components/PrettyInputText.js

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import prettyInputTextStyle from './css/prettyInputText.scss';

class PrettyInputText extends Component {

    constructor(props) {
        super(props);

        this.state = {
            labelShow: (this.props.inputValue.length > 0) ? true : false,
            inputActivated: false,
            errorActivated: false
        };

        this.handleTextInputChange = this.handleTextInputChange.bind(this);
        this.handleTextInputOnFocus = this.handleTextInputOnFocus.bind(this);
        this.handleTextInputOnBlur = this.handleTextInputOnBlur.bind(this);
        this.handleOnKeyPress = this.handleOnKeyPress.bind(this);
    }

    componentDidMount() {

    }

    static getDerivedStateFromProps(props, state) {
        if (props.inputValue.length > 0) {
            state = Object.assign(state, { labelShow: true });
        } else {
            state = Object.assign(state, { labelShow: false });
        }

        return state;
    }

    handleTextInputChange(e) {
        if (this.props.isEnabled) {
            if (e.target.value.toString().length > 0) {
                if (this.state.errorActivated) {
                    this.setState({ errorActivated: false });
                }
            } else {
                if (this.props.isRequired) {
                    this.setState({ errorActivated: true });
                }
            }
            if (this.props.onChange) {
                this.props.onChange(e);
            }
        }
    }

    handleTextInputOnFocus(e) {
        this.setState({ inputActivated: true });
    }

    handleTextInputOnBlur(e) {
        this.setState({ inputActivated: false });
        if (!this.props.onValidation(this.props.inputValue)) {
            this.setState({ errorActivated: true });
        }
    }

    handleOnKeyPress(e) {
        if (this.props.isEnabled) {
            if (this.props.onKeyPress) {
                this.props.onKeyPress(e);
            }
        }
    }

    render() {
        let mainContainerStyleClasses = prettyInputTextStyle['pretty-input-text'];
        if (this.state.errorActivated) {
            mainContainerStyleClasses += ' error';
        }
        let insideContainerStyleClasses = 'inside-container';
        if (this.state.inputActivated) {
            insideContainerStyleClasses += ' is-active';
        }
        if (this.state.labelShow) {
            insideContainerStyleClasses += ' show-label';
        }
        return (
            <div className={mainContainerStyleClasses} style={{width: `${this.props.width}px`}}>
                <div className={insideContainerStyleClasses} style={{background: this.props.backgroundColor}}>
                    <input
                        type="text"
                        maxLength={this.props.size}
                        name={this.props.name}
                        placeholder={this.props.labelText}
                        value={this.props.inputValue}
                        onChange={this.handleTextInputChange}
                        onFocus={this.handleTextInputOnFocus}
                        onBlur={this.handleTextInputOnBlur}
                        onKeyPress={this.handleOnKeyPress}
                        style={{color: this.props.labelColor}}
                        disabled={!this.props.isEnabled}
                    />
                    <label className="success-label" style={{color: this.state.labelShow ? this.props.labelColor: this.props.backgroundColor}}>{this.props.labelText}</label>
                    <label className="error-label">{this.props.errorValue}</label>

                </div>
            </div>
        )
    }
}

PrettyInputText.propTypes = {
    name: PropTypes.string.isRequired,
    labelText: PropTypes.string.isRequired,
    inputValue: PropTypes.string,
    errorValue: PropTypes.string,
    size: PropTypes.number,
    isEnabled: PropTypes.bool,
    isRequired: PropTypes.bool,
    onValidation: PropTypes.func,
    onChange: PropTypes.func,
    onKeyPress: PropTypes.func,
    labelColor: PropTypes.string,
    backgroundColor: PropTypes.string,
    width: PropTypes.number
};

PrettyInputText.defaultProps = {
    inputValue: '',
    errorValue: 'Error',
    size: 50,
    isEnabled: true,
    isRequired: false,
    onValidation: inputValue => true,
    onChange: null,
    onKeyPress: null,
    labelColor: '#0069ff',
    backgroundColor: '#fff',
    width: 200
};

export default PrettyInputText;
